libraries/standard/mqtt/src/mqtt.c

```diff
@@ -1391,58 +1515,60 @@ MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * const pContext,
 
-            if( status == MQTTSuccess )
-            {
-                /* Reset the status to indicate that we should not try to read
-                 * a packet from the transport interface. */
-                status = MQTTNoDataAvailable;
-            }
-        }
-        else if( status != MQTTSuccess )
+        if( status != MQTTSuccess )
         {
-            LogError( ( "Receiving incoming packet length failed. Status=%d",
-                        status ) );
+            LogError( ( "Exiting process loop. Error status=%d", status ) );
         }
-        else
+
+        /* Recalculate remaining time and check if loop should exit. */
+        elapsedTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
+
+        if( elapsedTimeMs > timeoutMs )
         {
-            /* Receive packet. Remaining time is recalculated at the end of the loop. */
-            status = receivePacket( pContext, incomingPacket, remainingTimeMs );
+            break;
         }
 
-        /* Handle received packet. If no data was read then this will not execute. */
-        if( status == MQTTSuccess )
-        {
-            incomingPacket.pRemainingData = pContext->networkBuffer.pBuffer;
+        remainingTimeMs = timeoutMs - elapsedTimeMs;
+    }
 
-            /* PUBLISH packets allow flags in the lower four bits. For other
-             * packet types, they are reserved. */
-            if( ( incomingPacket.type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
-            {
-                status = handleIncomingPublish( pContext, &incomingPacket );
-            }
-            else
-            {
-                status = handleIncomingAck( pContext, &incomingPacket );
-            }
-        }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+MQTTStatus_t MQTT_ReceiveLoop( MQTTContext_t * pContext,
+                               uint32_t timeoutMs )
+{
+    MQTTStatus_t status = MQTTBadParameter;
+    MQTTGetCurrentTimeFunc_t getTimeStampMs = NULL;
+    uint32_t entryTimeMs = 0U, remainingTimeMs = timeoutMs, elapsedTimeMs = 0U;
+    uint32_t adjustedTimeoutMs = timeoutMs;
 
-        if( status == MQTTNoDataAvailable )
+    if( pContext != NULL )
+    {
+        getTimeStampMs = pContext->callbacks.getTime;
+
+        if( getTimeStampMs == NULL )
         {
-            /* No data available is not an error. Reset to MQTTSuccess so the
-             * return code will indicate success. */
-            status = MQTTSuccess;
+            getTimeStampMs = getTimeStampNoSystemTime;
+            adjustedTimeoutMs = 0U;
+            remainingTimeMs = 0U;
         }
 
+        entryTimeMs = getTimeStampMs();
+        status = MQTTSuccess;
+    }
+    else
+    {
+        LogError( ( "MQTT Context cannot be NULL." ) );
+    }
+
+    while( status == MQTTSuccess )
+    {
+        status = receiveSingleIteration( pContext, remainingTimeMs, false );
+
         if( status != MQTTSuccess )
```

Comment:
Same comment as earlier, if it is send or receive failure loop should exit and error should be pass  to the application.

Reply:
See above. The loop exits, just at the end of the iteration instead of immediately

Reply:
Agree,  I saw that and realized that it must me MISRA. I was wondering if you can put else to make it more clear. Or may be just add comment.

Reply:
Added a comment